substrate/runtime/session/src/lib.rs

```diff
@@ -153,13 +156,16 @@ impl<T: Trait> Module<T> {
 		let block_number = <system::Module<T>>::block_number();
 		let is_final_block = ((block_number - Self::last_length_change()) % Self::length()).is_zero();
 		let broken_validation = Self::broken_validation();
-		if is_final_block || broken_validation {
-			Self::rotate_session(!broken_validation);
+		if let Some(normal_rotation) = Self::forcing_new_session() {
+			Self::rotate_session(normal_rotation, is_final_block);
```

Comment:
Not sure about how `force_new_session()` will be used, but probably `Self::rotate_session(normal_rotation && !broken_validation, is_final_block)` would be better (to make sure we slash validators)?